// @vue-macros/common@1.1.0/dist/ast.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});














var _chunkYADQPSAZjs = require('./chunk-YADQPSAZ.js');
require('./chunk-6FHO54WM.js');
require('./chunk-VBSW6ED5.js');















exports.TS_NODE_TYPES = _chunkYADQPSAZjs.TS_NODE_TYPES; exports.babelParse = _chunkYADQPSAZjs.babelParse; exports.checkInvalidScopeReference = _chunkYADQPSAZjs.checkInvalidScopeReference; exports.isCallOf = _chunkYADQPSAZjs.isCallOf; exports.isFunctionType = _chunkYADQPSAZjs.isFunctionType; exports.isLiteralType = _chunkYADQPSAZjs.isLiteralType; exports.isStaticExpression = _chunkYADQPSAZjs.isStaticExpression; exports.isStaticObjectKey = _chunkYADQPSAZjs.isStaticObjectKey; exports.resolveLiteral = _chunkYADQPSAZjs.resolveLiteral; exports.resolveObjectExpression = _chunkYADQPSAZjs.resolveObjectExpression; exports.resolveObjectKey = _chunkYADQPSAZjs.resolveObjectKey; exports.resolveTemplateLiteral = _chunkYADQPSAZjs.resolveTemplateLiteral; exports.unwrapTSNode = _chunkYADQPSAZjs.unwrapTSNode; exports.walkAST = _chunkYADQPSAZjs.walkAST;
exports.default = module.exports;
