// @vue-macros/common@1.1.0/dist/chunk-7J7RYQ3Y.mjs
import {
  babelParse
} from "./chunk-CNUWHFPH.mjs";
import {
  getLang
} from "./chunk-HFGYXSVL.mjs";
import {
  REGEX_VUE_SFC
} from "./chunk-7IYHK53L.mjs";

// src/vue.ts
import { parse } from "@vue/compiler-sfc";
function parseSFC(code, id) {
  var _a;
  const sfc = parse(code, {
    filename: id
  });
  const { descriptor, errors } = sfc;
  const lang = (_a = descriptor.script || descriptor.scriptSetup) == null ? void 0 : _a.lang;
  return {
    sfc,
    ...descriptor,
    lang,
    errors,
    getSetupAst() {
      if (!descriptor.scriptSetup)
        return;
      return babelParse(descriptor.scriptSetup.content, lang);
    },
    getScriptAst() {
      if (!descriptor.script)
        return;
      return babelParse(descriptor.script.content, lang);
    }
  };
}
function getFileCodeAndLang(code, id) {
  var _a;
  if (!REGEX_VUE_SFC.test(id))
    return { code, lang: getLang(id) };
  const sfc = parseSFC(code, id);
  const scriptCode = ((_a = sfc.script) == null ? void 0 : _a.content) ?? "";
  return {
    code: sfc.scriptSetup ? `${scriptCode}
;
${sfc.scriptSetup.content}` : scriptCode,
    lang: sfc.lang ?? "js"
  };
}
function addNormalScript({ script, lang }, s) {
  return {
    start() {
      if (script)
        return script.loc.end.offset;
      const attrs = lang ? ` lang="${lang}"` : "";
      s.prependLeft(0, `<script${attrs}>`);
      return 0;
    },
    end() {
      if (!script)
        s.appendRight(0, `
</script>
`);
    }
  };
}

export {
  parseSFC,
  getFileCodeAndLang,
  addNormalScript
};
