// @vue-macros/common@1.1.0/dist/chunk-7IYHK53L.mjs
// src/constants.ts
var DEFINE_PROPS = "defineProps";
var DEFINE_PROPS_DOLLAR = "$defineProps";
var DEFINE_PROPS_REFS = "definePropsRefs";
var DEFINE_EMITS = "defineEmits";
var WITH_DEFAULTS = "withDefaults";
var DEFINE_OPTIONS = "defineOptions";
var DEFINE_MODEL = "defineModel";
var DEFINE_MODEL_DOLLAR = "$defineModel";
var DEFINE_SETUP_COMPONENT = "defineSetupComponent";
var DEFINE_RENDER = "defineRender";
var DEFINE_SLOTS = "defineSlots";
var REPO_ISSUE_URL = "https://github.com/sxzz/unplugin-vue-macros/issues";
var REGEX_SRC_FILE = /\.[cm]?[jt]sx?$/;
var REGEX_SETUP_SFC = /\.setup\.[cm]?[jt]sx?$/;
var REGEX_VUE_SFC = /\.vue$/;
var REGEX_VUE_SUB = /\.vue\?vue&/;
var REGEX_DTS = /\.d\.[cm]?ts$/;
var REGEX_NODE_MODULES = /node_modules/;
var REGEX_LANG_TS = /^[cm]?tsx?$/;
var REGEX_LANG_JSX = /^[cm]?[jt]sx$/;

export {
  DEFINE_PROPS,
  DEFINE_PROPS_DOLLAR,
  DEFINE_PROPS_REFS,
  DEFINE_EMITS,
  WITH_DEFAULTS,
  DEFINE_OPTIONS,
  DEFINE_MODEL,
  DEFINE_MODEL_DOLLAR,
  DEFINE_SETUP_COMPONENT,
  DEFINE_RENDER,
  DEFINE_SLOTS,
  REPO_ISSUE_URL,
  REGEX_SRC_FILE,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  REGEX_DTS,
  REGEX_NODE_MODULES,
  REGEX_LANG_TS,
  REGEX_LANG_JSX
};
