// @vue-macros/common@1.1.0/dist/constants.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});




















var _chunkVBSW6ED5js = require('./chunk-VBSW6ED5.js');





















exports.DEFINE_EMITS = _chunkVBSW6ED5js.DEFINE_EMITS; exports.DEFINE_MODEL = _chunkVBSW6ED5js.DEFINE_MODEL; exports.DEFINE_MODEL_DOLLAR = _chunkVBSW6ED5js.DEFINE_MODEL_DOLLAR; exports.DEFINE_OPTIONS = _chunkVBSW6ED5js.DEFINE_OPTIONS; exports.DEFINE_PROPS = _chunkVBSW6ED5js.DEFINE_PROPS; exports.DEFINE_PROPS_DOLLAR = _chunkVBSW6ED5js.DEFINE_PROPS_DOLLAR; exports.DEFINE_PROPS_REFS = _chunkVBSW6ED5js.DEFINE_PROPS_REFS; exports.DEFINE_RENDER = _chunkVBSW6ED5js.DEFINE_RENDER; exports.DEFINE_SETUP_COMPONENT = _chunkVBSW6ED5js.DEFINE_SETUP_COMPONENT; exports.DEFINE_SLOTS = _chunkVBSW6ED5js.DEFINE_SLOTS; exports.REGEX_DTS = _chunkVBSW6ED5js.REGEX_DTS; exports.REGEX_LANG_JSX = _chunkVBSW6ED5js.REGEX_LANG_JSX; exports.REGEX_LANG_TS = _chunkVBSW6ED5js.REGEX_LANG_TS; exports.REGEX_NODE_MODULES = _chunkVBSW6ED5js.REGEX_NODE_MODULES; exports.REGEX_SETUP_SFC = _chunkVBSW6ED5js.REGEX_SETUP_SFC; exports.REGEX_SRC_FILE = _chunkVBSW6ED5js.REGEX_SRC_FILE; exports.REGEX_VUE_SFC = _chunkVBSW6ED5js.REGEX_VUE_SFC; exports.REGEX_VUE_SUB = _chunkVBSW6ED5js.REGEX_VUE_SUB; exports.REPO_ISSUE_URL = _chunkVBSW6ED5js.REPO_ISSUE_URL; exports.WITH_DEFAULTS = _chunkVBSW6ED5js.WITH_DEFAULTS;
exports.default = module.exports;
