// vscode-vue-languageservice/out/services/referencesCodeLens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
function register({ sourceFiles }) {
    return (uri) => {
        const sourceFile = sourceFiles.get(uri);
        if (!sourceFile)
            return [];
        const document = sourceFile.getTextDocument();
        const result = [];
        for (const sourceMap of sourceFile.getTsSourceMaps()) {
            for (const maped of sourceMap.mappings) {
                if (!maped.data.capabilities.referencesCodeLens)
                    continue;
                const data = {
                    lsType: sourceMap.lsType,
                    uri: uri,
                    offset: maped.sourceRange.start,
                    tsUri: sourceMap.mappedDocument.uri,
                    tsOffset: maped.mappedRange.start,
                };
                result.push({
                    range: {
                        start: document.positionAt(maped.sourceRange.start),
                        end: document.positionAt(maped.sourceRange.end),
                    },
                    // @ts-expect-error
                    data,
                });
            }
        }
        return result;
    };
}
exports.register = register;
