// @vue-macros/common/dist/chunk-D7U5WAVG.mjs
import {
  babelParse
} from "./chunk-CNUWHFPH.mjs";
import {
  getLang
} from "./chunk-HFGYXSVL.mjs";
import {
  REGEX_VUE_SFC
} from "./chunk-7IYHK53L.mjs";

// src/vue.ts
import { parse } from "@vue/compiler-sfc";
function parseSFC(code, id) {
  var _a;
  const sfc = parse(code, {
    filename: id
  });
  const { descriptor, errors } = sfc;
  const lang = (_a = descriptor.script || descriptor.scriptSetup) == null ? void 0 : _a.lang;
  return {
    sfc,
    ...descriptor,
    lang,
    errors,
    getSetupAst() {
      if (!descriptor.scriptSetup)
        return;
      return babelParse(descriptor.scriptSetup.content, lang);
    },
    getScriptAst() {
      if (!descriptor.script)
        return;
      return babelParse(descriptor.script.content, lang);
    }
  };
}
function getFileCodeAndLang(code, id) {
  var _a;
  if (!REGEX_VUE_SFC.test(id))
    return { code, lang: getLang(id) };
  const sfc = parseSFC(code, id);
  const scriptCode = ((_a = sfc.script) == null ? void 0 : _a.content) ?? "";
  return {
    code: sfc.scriptSetup ? `${scriptCode}
;
${sfc.scriptSetup.content}` : scriptCode,
    lang: sfc.lang ?? "js"
  };
}
function addNormalScript({ script, lang }, s) {
  return {
    start() {
      if (script)
        return script.loc.end.offset;
      const attrs = lang ? ` lang="${lang}"` : "";
      s.prependLeft(0, `<script${attrs}>`);
      return 0;
    },
    end() {
      if (!script)
        s.appendRight(0, `
</script>
`);
    }
  };
}
var imported = /* @__PURE__ */ new WeakMap();
var HELPER_PREFIX = "__MACROS_";
function importHelperFn(s, offset, name, from) {
  var _a;
  const cacheKey = `${from}@${name}`;
  if (!((_a = imported.get(s)) == null ? void 0 : _a.has(cacheKey))) {
    s.appendLeft(
      offset,
      `
import { ${name} as ${HELPER_PREFIX}${name} } from '${from}';`
    );
    if (!imported.has(s)) {
      imported.set(s, /* @__PURE__ */ new Set([cacheKey]));
    } else {
      imported.get(s).add(cacheKey);
    }
  }
}

export {
  parseSFC,
  getFileCodeAndLang,
  addNormalScript,
  HELPER_PREFIX,
  importHelperFn
};
